### PR TITLE
Furina Implementation

### PR DESCRIPTION
**Pull request by Charlie-Zheng**
* [x]  New character package
* [x]  Config in character package
* [x]  Run pipeline with added config
* [x]  Character key
* [x]  Shortcuts for character key
* [x]  Add Character package to imports
* [x]  Character talent stats
* [x]  Character curve
* [x]  Normal Attack
* [x]  Charge Attack / Aimed Shot
* [x]  Skill
* [x]  Burst
* [x]  A1
* [x]  A4
* [x]  C0
* [x]  C1
* [x]  C2
* [x]  C3
* [x]  C4
* [x]  C5
* [x]  C6
* [x]  Other necessary talents (custom dash/jump, low/high plunge, ...)
* [x]  Hitlag
* [x]  ICD
* [x]  StrikeType
* [x]  Hitboxes
* [x]  Attack durability
* [x]  Particles
* [x]  Frames
* [x]  Update documentation (Add charname.md)
* [x] Update localization (IngameNames.json)

### CODE REVIEW

**Comment by k0l11** [comment]
Missing a furina.md

File: ui/packages/docs/src/components/AoE/character_data.json [hunk 1]
```diff
@@ -2348,6 +2348,115 @@
...
+      },
+      {
+        "ability": "N4-C6",
+        "shape": "Box",
+        "center": "PrimaryTarget",
+        "offsetY": -3,
+        "boxX": 5,
+        "boxY": 6
+      }
+    ],
```
Comment (k0l11) [inline ui/packages/docs/src/components/AoE/character_data.json:2413]: Remove the offset for now following the comment in attack.go.

File: internal/characters/furina/charge.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			}
+		}, chargeHitmark+1)
+	} else {
+		c.QueueCharTask(func() {
+			c.arkhe = ousia
+			if c.StatusIsActive(skillKey) {
+				c.summonSalonMembers(0)
+			}
+		}, chargeHitmark+1)
+	}
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/charge.go:73]: This the `c.QueueCharTask(arkheChangeFunc, chargeHitmark)` guarantee that the arkhe switch happens after the CA does damage? This is to make sure the c6 damage type/buff type is correct

File: internal/characters/furina/cons.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		if !c.StatusIsActive(c6OusiaHealKey) {
+			c.c6HealSrc = c.Core.F
+			for _, char := range c.Core.Player.Chars() {
+				char.QueueCharTask(c.c6heal(char, c.Core.F), 60)
+				char.AddStatus(c6OusiaHealKey, 2.9*60, true)
+			}
+		} else {
+			for _, char := range c.Core.Player.Chars() {
+				char.ExtendStatus(c6OusiaHealKey, 2.9*60)
+			}
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/cons.go:81]: I saw a vod with c6 furina. Her C6 actually extends duration, so if you do E N6 her healing lasts 17s+. Her C6 doesn't refresh duration, but actually extends it.

[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/cons.go:81]: Let me find the video. I believe it showed that it was affected by hitlag per character too

[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/cons.go:81]: https://www.youtube.com/watch?v=4aWLfPzA1WU

File: internal/characters/furina/charge.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/furina/charge.go:73]: I'm not sure, pls check if chargeHitmark is enough to achieve that

File: internal/characters/furina/attack.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	"github.com/genshinsim/gcsim/pkg/core/geometry"
+)
+
+var (
+	attackFrames    [][]int
+	attackHitmarks  = []int{15, 12, 21, 27}
+	attackOffsets   = []float64{-1.4, -0.85, -0.95, -3}
+	attackOffsetsC6 = []float64{-1.5, -1.15, -1.1, -3}
+
+	// these ones should be correct
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/attack.go:20]: The hitlag frames and hitbox sizes are from homdgcat, should I still leave it as a todo?

[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/furina/attack.go:20]: Nope.

File: internal/characters/furina/burst.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			})
+
+			char.AddHealBonusMod(character.HealBonusMod{
+				Base: modifier.NewBase("furina-burst-heal-buff", burstDur),
+				Amount: func() (float64, bool) {
+					return common.Min(c.curFanfare, c.maxQFanfare) * burstFanfareHBRatio[c.TalentLvlBurst()], false
+				},
+			})
+		}
+	}, burstHitmark+1)
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/burst.go:140]: That's earlier than the burstHitmark, that means the damage should benefit from C1. I just checked in game and confirmed that the burst hit benefits from the C1 fanfare.

File: internal/characters/furina/burst.go [hunk 6]
```diff
@@ -37,7 +36,7 @@ func (c *char) addFanfare(amt float64) {
 	if c.Base.Cons >= 2 {
 		amt *= 3.5
 	}
-	c.curFanfare = common.Min(c.maxFanfare, c.curFanfare+amt)
+	c.curFanfare = math.Min(c.maxFanfare, c.curFanfare+amt)
```
Comment (k0l11) [inline internal/characters/furina/burst.go]: You were supposed to use the builtin min/max, not math.Min/Max. Please adjust that for Furina in all files.

File: internal/characters/furina/burst.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/furina/burst.go]: Adding a comment that explains this interaction would be nice.

File: internal/characters/furina/burst.go [hunk 7]
```diff
@@ -95,6 +94,32 @@ func (c *char) burstInit() {
...
+	for _, char := range c.Core.Player.Chars() {
+		char.AddAttackMod(character.AttackMod{
+			Base: modifier.NewBase("furina-burst-damage-buff", -1),
+			Amount: func(atk *combat.AttackEvent, t combat.Target) ([]float64, bool) {
+				if c.StatusIsActive(burstKey) {
+					c.burstBuff[attributes.DmgP] = math.Min(c.curFanfare, c.maxQFanfare) * burstDMGRatio
+				} else {
+					c.burstBuff[attributes.DmgP] = 0
+				}
+				return c.burstBuff, true
```
Comment (k0l11) [inline internal/characters/furina/burst.go:109]:
```
if !c.StatusIsActive(burstKey) {
	return nil, false 
} 
c.burstBuff[attributes.DmgP] = math.Min(c.curFanfare, c.maxQFanfare) * burstDMGRatio
return c.burstBuff, true
```

File: internal/characters/furina/charge.go [hunk 8]
```diff
@@ -28,53 +29,58 @@ func init() {
...
 				c.summonSalonMembers(0)
 			}
-		}, chargeHitmark+1)
+		}
 	}
+	// +1 so that c6 evaluates DMG bonus/Heal status correctly
+	c.QueueCharTask(arkheChangeFunc, chargeHitmark-windup+1)
+
 	return action.Info{
 		Frames:          frames.NewAbilFunc(chargeFrames),
```
Comment (k0l11) [inline internal/characters/furina/charge.go:81]:
```
Frames:          func(next action.Action) int { return chargeFrames[next] - windup },
```

File: internal/characters/furina/skill.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+	singerInterval = 120
+)
+
+func (c *char) calcSalonTick(tickNum int, initialTick, interval float64) int {
+	// the distribution is left skewed. We approxiamated with boxcox with lambda 0.728
+	// then used the transformation to convert from norm dist to the experimental distribution
+	randOffset := math.Pow(common.Max(c.Core.Rand.NormFloat64()*1.0403+4.073023273, 0.0), (1/0.728)) - 7
+
+	// this limits the offset to [-7, 7]
+	randOffset = common.Min(randOffset, 7)
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/skill.go]: I tested with a short python script to generate and plot the frequency distribution.
```python
import numpy as np
import matplotlib.pyplot as plt
data = np.random.normal(0, 1, 100000)
data = [max(d*1.0403+4.0730, 0)**(1/0.728)-7 for d in data]
plt.hist(data, bins=50, density=True, alpha=0.6, color='b')
plt.show()
```
image
image

File: internal/characters/furina/skill.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+			return
+		}
+
+		if !c.StatusIsActive(skillKey) {
+			return
+		}
+
+		c.Core.QueueAttack(ai, ap, 0, 0, callbacks...)
+	}, delay)
+}
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/furina/skill.go:178]: I'm actually not sure if the projectiles disappear if the summon src changes due to CA. I'll add a comment explaining why I'm coding it this way, but also add a TODO so someone can check if this is correct.

[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/furina/skill.go:178]: Sounds good.

File: internal/characters/furina/burst.go (hunk 6 above)
Comment (Charlie-Zheng) [reply internal/characters/furina/burst.go]: Can I do it for all the Golang files or just Furina?

Comment (k0l11) [reply internal/characters/furina/burst.go]: Just Furina pls.

**Pull request merged**